vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+	keys, err := req.Storage.List(ctx, namedKeyConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(keys), nil
+}
+
+// pathOIDCRotateKey is used to manually trigger a rotation on the named key
+func (i *IdentityStore) pathOIDCRotateKey(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	// load the named key and perform a rotation
+	entry, err := req.Storage.Get(ctx, namedKeyConfigPath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return logical.ErrorResponse("no named key found at %q", name), logical.ErrInvalidRequest
+	}
+
+	var storedNamedKey namedKey
+	if err := entry.DecodeJSON(&storedNamedKey); err != nil {
+		return nil, err
+	}
+
+	// call rotate with an appropriate overrideTTL where < 0 means no override
+	verificationTTLOverride := -1 * time.Second
+
+	if ttlRaw, ok := d.GetOk("verification_ttl"); ok {
+		verificationTTLOverride = time.Duration(ttlRaw.(int)) * time.Second
+	}
+
+	if err := storedNamedKey.rotate(ctx, req.Storage, verificationTTLOverride); err != nil {
+		return nil, err
+	}
+
+	i.oidcCache.purge()
+
+	return nil, nil
+}
+
+func (i *IdentityStore) pathOIDCKeyExistenceCheck(ctx context.Context, req *logical.Request, d *framework.FieldData) (bool, error) {
+	name := d.Get("name").(string)
+
+	entry, err := req.Storage.Get(ctx, namedKeyConfigPath+name)
+	if err != nil {
+		return false, err
+	}
+
+	return entry != nil, nil
+}
+
+// handleOIDCGenerateSignToken generates and signs an OIDC token
+func (i *IdentityStore) pathOIDCGenerateToken(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	roleName := d.Get("name").(string)
+
+	role, err := i.getOIDCRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse("role %q not found", roleName), nil
+
+	}
+
+	entry, _ := req.Storage.Get(ctx, namedKeyConfigPath+role.Key)
+	if entry == nil {
+		return logical.ErrorResponse("key %q not found", role.Key), nil
+	}
+
+	var key namedKey
+	if err := entry.DecodeJSON(&key); err != nil {
+		return nil, err
+	}
+
+	// generate an OIDC token from entity data
+	accessorEntry, err := i.core.tokenStore.lookupByAccessor(ctx, req.ClientTokenAccessor, false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	te, err := i.core.LookupToken(ctx, accessorEntry.TokenID)
+	if err != nil {
+		return nil, err
+	}
+
+	if te == nil || te.EntityID == "" {
+		return logical.ErrorResponse("No entity associated with this Vault token"), nil
+	}
+
+	config, err := i.getOIDCConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now()
+	idToken := idToken{
+		Issuer:   config.effectiveIssuer,
+		Subject:  te.EntityID,
+		Audience: role.RoleID,
+		Expiry:   now.Add(role.TokenTTL).Unix(),
```

Comment:
A bit pedantic, but, rather than call time.Now multiple times, can you call it once and re-use it? Technically if the two calls happen over a second boundary you could end up with an expiry that is one second off from what it ought to be.

Reply:
I did find a couple cases of multiple `time.Now`s elsewhere that I've just fixed, but that's not the case here, where we're reusing it as `now` in Expiry and IssuedAt. Or I'm misunderstanding.

Reply:
I think it was a case of me jumping around, I saw that elsewhere and clicked on somewhere (wrong) that was using times. No worries!